pennylane/tape/qscript.py

```diff
@@ -299,6 +300,12 @@ def measurements(self):
         """
         return self._measurements
 
+    @property
+    @functools.lru_cache()
```

Comment:
Would be good to have a test which checks if this is being cached properly. I think this is an instance based cache, so we should be fine, but I am not sure how it acts if we query this property and then update the tape.

Reply:
This is a great catch! I've added a test that fails if caching is used.

Therefore, I would resort to not using `lru_cache` because we are updating the `_measurements` attribute in the tape expansion and other spots. The caching we desire requires more machinery that is out of scope here and would not result in extreme savings. Computing the `tape.samples_computational_basis` boolean for the worst-case scenario where the computational basis sampling is the last measurement when having ~100000 measurements would mean going from ~25 ms to ~75 ns for
```py
measurements = [qml.expval(qml.PauliZ(w)) for w in range(100000)]
measurements.append(qml.sample())

tape = qml.tape.QuantumTape([], measurements)
```

Reply:
Awesome!